### DISCUSSION TITLE
Paired or Unpaired

### DISCUSSION THREAD

**Discussion by John1231983** [voxelmorph/voxelmorph#99] [Q&A] [answered]
I would like to ask you a question about my dataset. I have a dataset that has 10 paired data and 100 unpaired data. They are affined transform to the atlas. Which option do you prefer to train voxelmorph?

1. Use only 10 paired data that allows the network easy to train

2. Use 100 unpaired data that enable the network to make a better generalization

In the testing, the performance is evaluated on 15 paired data.

**Comment by adalca** [comment]
What does 'paired data' mean here? T1 to T2? Or two T1s where you have the ground truth registration? or...?

**Comment by John1231983** [comment]
Paired means the CT and MRI that captures from same subject

**Comment by adalca** [comment]
So your goal is to train a CT-to-MRI registration network? To be used in paired situations (within same subject?).

**Comment by John1231983** [comment]
Yes. That right. I am having two options to prepare dataset. So, I ask you about that. If I used paired data, it looks fine but the number of paired data is too small. It may be overfitting or not good generalization

**Comment by adalca** [comment, marked as answer]
I'm not sure. I suspect using the 10 paired subjects (combined with some spatial-deformation data augmentation) would do the job, actually, since the deformations (after rigid alignment) would be tiny, and so would not over-fit too much. 

The problem with using the unpaired is that the deformations fields will be *much* bigger, so they won't be a good representation of within-subject registration (which is your test goal)

**Comment by John1231983** [comment]
What is some spatial-deformation data augmentation mean? Thanks

**Comment by adalca** [comment]
Applying random (smooth) deformations to your input, as data augmentation. But actually, perhaps it doesn't make as much sense here since the true deformations are very small. I would just start by trying the 10 paired subjects.

**Comment by John1231983** [comment]
Have you try it in voxelmorph? Does it improve performance?

**Comment by adalca** [comment]
I believe so when you have very few subjects. But of course you have to simulate reasonable deformation fields!

We've mostly used it in the context of segmentation, see https://arxiv.org/abs/1902.09383

**Comment by John1231983** [comment]
Thanks. I got it. Just small question. If I consider the subject-to-subject registration. And each subject has a paired CT+MRI (already aligned). What is the option do you prefer for registration:

1. Register MRI of a subject to another subject and obtain the deformation field. Warp the CT with the deformation to obtain warp CT. This way only used MRI as input

2. Concatenate MRI and CT together as 2 channel and performance registration simultaneously. Input will be `BxDxHxWx2` 

3. Register MRI of a subject to other subject, called this is loss1. Register CT to CT, called this is loss 2. The total loss is the summation of loss 1 and loss 2.

**Comment by adalca** [comment]
I would try option 2, making sure that both images are normalized!

**Comment by John1231983** [comment]
Thanks. Any reason why image should normalize to [0 1] instead of another normalize likes zero mean unit variance or [-1 1]. I guess that is some requirements for ncc loss

**Comment by adalca** [comment]
no, i think other normalizations are fine as well. 

LNCC loss does neighbourhood normalization anyway, so I think anything goes.

**Comment by John1231983** [comment]
@adalca : I tried three options and only the first option works. I have no reason. In additions, only the [0,1] range works well, while the [-1,1] gives a positive value of ncc loss. I guess reason is that I used zero padding to compute ncc loss (zero padding on background region). While the range [-1,1] background region has value -1, but we still use zero padding for the region

**Comment by adalca** [comment]
@John1231983 
Thanks. I expected the first one to work, but I was hoping the second would be better. Any ideas why it failed?

yes, if you are padding it should be the same as the background!

**Comment by John1231983** [comment]
Thanks. I have no idea. I tried with another dataset that also subject to subject registration but using t1+t2. It means t1 register to t1 and t2 register to t2. I found that only t1 to t1 works, while t2 to t2 has loss very unstable, sometine it is -0.2, but sometimes 1.8 or 233.0. So I guess is that second modality has the problem

**Comment by adalca** [comment]
We've definately registered T2s without an issue. Are you sure they were in [0,1]? also, what loss were you using?

**Comment by John1231983** [comment]
I am using ncc loss and intensities normalized to [0,1]. Yes. I think if it worked in t1 then it should work in other modalities.  The smooth weight lambda is 1.0

**Comment by adalca** [comment]
can you try MSE just to help debug why NCC blows up?

**Comment by longkuanzou** [comment]
> Paired means the CT and MRI that captures from same subject

Hi, do you mean the CT image was registrated by MRI image?

**Comment by John1231983** [comment]
Yes. That is paired images but must same subjects. If they come from different subjects the registration error will be larger

**Comment by longkuanzou** [comment]
> Yes. That is paired images but must same subjects. If they come from different subjects the registration error will be larger

Thanks for your reply.
 I' m new in image registration. and what do you mean "same subjects"? sorry for asking such a stupid question.

By the way, In your trying [https://github.com/voxelmorph/voxelmorph/issues/93] also got some questions.
1/  In your first step, how do you "linear rigid register" each subject to a template? Is there have useful library (like SimpleITK, or ANTs) to mask this? 
2/ I found that, the data in your link [https://www.synapse.org/#!Synapse:syn3281031](url) seems  without skull, so the data is preprocessed or originnal data?
3/ And can you provide all the registered data of OAIS? I want to trian the voxelmorph model, but haven't paired data.
